### DISCUSSION TITLE
Docker + ASO + Env Vars

### DISCUSSION THREAD

**Discussion by mAAdhaTTah** [vercel/next.js#16213] [Help]
We're working on migrating a couple of things at once: full stack Django app -> Django API + Next.js, and standard "code on the box" style -> Docker (EC2 -> Fargate/ECS). We're _also_ integrating w/ a headless CMS. This means that we have a few areas of the code that would need env vars:

1. `getStaticPaths/Props` on the catch-all route used to serve CMS pages.
    * This is because the test envs need to server draft content as well.
2. `REWRITE_TARGET` on the rewrites.
    * This is because each env has a separate Django app associated with it.

Ideally, with Docker, you build one container and deploy it to multiple environments, but these env vars are involved at buildtime rather than runtime.

What options do I have here? Do any of them allow me to continue to take advantage of ASO?

**Comment by c0b41** [comment]
you can pass env variables with docker cli https://docs.docker.com/engine/reference/run/#env-environment-variables or docker compose

**Reply by mAAdhaTTah** [reply]
This issue isn't the mechanics of how to pass env vars but how to set up a reproducible build that can be deployed across environments despite the necessity of env vars during build.

**Comment by rafaelalmeidatk** [comment]
> What options do I have here? Do any of them allow me to continue to take advantage of ASO?

The problem isn't with ASO but with static pages. If you want to be able to create a single Docker image and use it with different environment variables, you won't be able to build static pages at all. The value of the variables is going to change so you need to use SSR (`getServerSideProps`) for all the pages that are going to use these environment variables.

If you want to have the benefit of static pages, you would have to build different images for each environment you plan to deploy your app

**Reply by mAAdhaTTah** [reply]
Ah, got my terminology mixed up. Curious your thoughts on [this solution](https://github.com/vercel/next.js/discussions/11552#discussioncomment-33888): setting `paths = []` during the Docker build and using `revalidate` to generate the pages at runtime instead.

**Reply by rafaelalmeidatk** [reply]
Yeah that might work! But I am not 100% sure if the [Public Runtime Config](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) will work in the pages with `getStaticProps`, you may want to verify it
